File: online_store/domain/factory/product_factory.py
from online_store.domain.models.products.Book import Book
from online_store.domain.models.products.ProductInterface import ProductInterface
from online_store.domain.models.products.Electronic import Electronic
    
class ProductFactory:
    def create_product(self, product_type, *args, **kwargs) -> ProductInterface:
        if product_type == "Book":
            return Book(*args, **kwargs)
        elif product_type == "Electronic":
            return Electronic(*args, **kwargs)

File: tmpsjava/online_shop/src/domain/factory/ProductFactory.java
package domain.factory;

import domain.models.products.Book;
import domain.models.products.Electronic;
import domain.models.products.ProductInterface;
import domain.models.products.Promotion;

public class ProductFactory {
    public ProductInterface createProduct(String productType, Object... args) {
        return switch (productType) {
            case "Book" -> this.createBook(args);
            case "Electronic" -> this.createElectronic(args);
            // Handle other product types if needed
            default -> throw new IllegalArgumentException("Unknown product type: " + productType);
        };
    }

    private ProductInterface createBook(Object[] args) {
        // Extract arguments specific to Book, assuming title and price
        String title = (String) args[0];
        float price = (Float) args[1];
        Promotion promotion = (Promotion) args[2];
        // Add other Book-specific arguments as needed

        return (ProductInterface) new Book(title, price, promotion);
    }

    private ProductInterface createElectronic(Object[] args) {
        // Extract arguments specific to Electronic, assuming model and price
        String model = (String) args[0];
        float price = (Float) args[1];
        Promotion promotion = (Promotion) args[2];
        // Add other Electronic-specific arguments as needed

        return (ProductInterface) new Electronic(model, price, promotion);
    }
}